Std/Data/Array/Lemmas.lean

```diff
@@ -724,3 +724,182 @@ theorem all_eq_true_iff_forall_mem {l : Array α} : l.all p ↔ ∀ x, x ∈ l 
 /-! ### erase -/
 
 @[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+
+/-! ### swap -/
+
+@[simp] theorem swap_def_jeqi {a: Array α} {i j : Nat}
+    (_ : i < a.size := by trivial) (_ : j < a.size := by trivial)
+    (_ : j < (a.swap ⟨i, ‹_›⟩ ⟨j, ‹_›⟩).size := by trivial) : (a.swap ⟨i, _⟩ ⟨j, _⟩)[j] = a[i] :=
+  let i : Fin a.size := ⟨i, _⟩
+  let j : Fin a.size := ⟨j, _⟩
+
+  show (a.swap i j)[j] = a[i] from aux
+where aux {i j : Fin a.size} : (a.swap i j)[j] = a[i] := by simp [swap, fin_cast_val]
+
+@[simp] theorem swap_def_ieqj {a: Array α} {i j : Nat}
+    (_ : i < a.size := by trivial) (_ : j < a.size := by trivial)
+    (_ : i < (a.swap ⟨i, ‹_›⟩ ⟨j, ‹_›⟩).size := by trivial) : (a.swap ⟨i, _⟩ ⟨j, _⟩)[i] = a[j] :=
+  let i : Fin a.size := ⟨i, _⟩
+  let j : Fin a.size := ⟨j, _⟩
+
+  show (a.swap i j)[i] = a[j] from aux
+where aux {i j : Fin a.size} : (a.swap i j)[i] = a[j] :=
+  let vi := a.get i
+  let vj := a.get j
+  let j := a.size_set i vj ▸ j
+
+  let a := a.set i vj
+  let a' := a.set j vi
+
+  suffices a'[i.val]'(by simp [a, a']) = vj from by simp [a'] at this; assumption
+  if he : j.val = i.val then
+    by simp [←he, vi, vj, j, a', fin_cast_val]
+  else
+    suffices a[i.val]'(by simp [a]) = vj from (a.get_set_ne j vi _ he).symm ▸ this
+    by simp [get_set_ne, a]
+
+@[simp] theorem swap_def_else {a: Array α} {i j : Fin a.size}
+    (_ : p < a.size := by trivial) (_ : p < (a.swap i j).size := by trivial)
+    (_ : i.val ≠ p) (_ : j.val ≠ p) : (a.swap i j)[p] = a[p] :=
+  let i : Fin a.size := ⟨i, _⟩
+  let j : Fin a.size := ⟨j, _⟩
+  let p : Fin a.size := ⟨p, ‹_›⟩
+
+  show (a.swap i j)[p] = a[p] by apply aux <;> simp_all only [ne_eq, not_false_eq_true]
+where aux {i j p: Fin a.size} (hi : i.val ≠ p.val) (hj : j.val ≠ p.val) : (a.swap i j)[p] = a[p] :=
+  let vi := a.get i
+  let vj := a.get j
+
+  let j' := a.size_set i vj ▸ j
+  let a_ := a.set i vj
+  let a' := a_.set j' vi
+
+  suffices a'[p.val]'(by simp [a_, a']) = a[p.val] by simp [a'] at this; assumption
+
+  have : j.val=j'.val := by simp [j', fin_cast_val]
+  have h1 : a'[p.val]'(by simp [a', a_]) = a_[p.val]'(by simp [a_]) :=
+    a_.get_set_ne (j') vi _ (this ▸ hj )
+  have h2 : a_[p.val]'(by simp [a_]) = a[p.val] := a.get_set_ne i vj _ hi
+
+  trans h1 h2
+
+
+theorem swap_p {a: Array α} {i j p : Fin a.size} :
+    (a.swap i j)[p] = a[if i.val=p.val then j else if j.val=p.val then i else p] :=
```

Comment:
You can just write `i = p`, no need for `i.val = p.val`.

Reply:
@semorrison I have re-written the proofs of `swap_swap` and `swap_comm` from scratch to make them shorter and easier to follow. In particular they now nolonger rely on these two theorems `swap_p` and `swap_p'` ( aka.. `get_swap_eq_get_if`/`get_swap_if`) whose only purpose in my orignial draft was as prerequisite lemmas for `swap_swap` and `swap_comm`.. Since I now prove the essential theorems `swap_swap`/`swap_comm` without `swap_p` /`swap_p'` (and done so with a more concise shorter proof), I have simply removed  `swap_p` and `swap_p'` from the scope of this PR